storages/backends/azure_storage.py

```diff
@@ -1,42 +1,83 @@
 import mimetypes
 import os.path
-import time
-from datetime import datetime
-from time import mktime
+from datetime import datetime, timedelta
+from tempfile import NamedTemporaryFile
 
-from django.core.exceptions import ImproperlyConfigured
-from django.core.files.base import ContentFile
+from azure.common import AzureMissingResourceHttpError
+from azure.storage import CloudStorageAccount
+from azure.storage.blob import ContentSettings
+from django.core.files.base import File
 from django.core.files.storage import Storage
 from django.utils.deconstruct import deconstructible
 
 from storages.utils import setting
 
-try:
-    import azure  # noqa
-except ImportError:
-    raise ImproperlyConfigured(
-        "Could not load Azure bindings. "
-        "See https://github.com/WindowsAzure/azure-sdk-for-python")
-
-try:
-    # azure-storage 0.20.0
-    from azure.storage.blob.blobservice import BlobService
-    from azure.common import AzureMissingResourceHttpError
-except ImportError:
-    from azure.storage import BlobService
-    from azure import WindowsAzureMissingResourceError as AzureMissingResourceHttpError
-
 
 def clean_name(name):
     return os.path.normpath(name).replace("\\", "/")
 
 
+def pad_left(n, width, pad="0"):
+    return ((pad * width) + str(n))[-width:]
+
+
+@deconstructible
+class AzureStorageFile(File):
+
+    def __init__(self, name, mode, storage):
+        self._name = name
+        self._mode = mode
+        self._storage = storage
+        self._is_dirty = False
+        self._file = None
+
+    def _get_file(self):
+        if self._file is None:
+            self._file = NamedTemporaryFile(
+                suffix=".AzureBoto3StorageFile",
+                dir=setting("FILE_UPLOAD_TEMP_DIR", None)
+            )
+            if 'r' in self._mode:
+                self._is_dirty = False
+                self._storage.connection.get_blob_to_path(container_name=self._storage.azure_container,
+                                                          blob_name=self._name, file_path=self._file.name,
+                                                          max_connections=setting("AZURE_READ_MAX_CONNECTIONS", 2))
+
+                self._file.seek(0)
+        return self._file
+
+    file = property(_get_file)
+
+    def read(self, *args, **kwargs):
+        if 'r' not in self._mode:
+            raise AttributeError("File was not opened in read mode.")
+        return super(AzureStorageFile, self).read(*args, **kwargs)
+
+    def write(self, content):
+        if 'w' not in self._mode:
+            raise AttributeError("File was not opened in write mode.")
+        self._is_dirty = True
+        ret = super(AzureStorageFile, self).write(content)
+        return ret
+
+    def close(self):
```

Comment:
Can somebody tell me what the benefit is of writing the file first to local storage and then to Azure inside the close method? I see only more moving parts which can break, I would prefer to write directly to Azure.

Reply:
To upload a new file, you'll do `default_storage.save(some_file_or_stream)`, that's what django does behind the scenes. The `File` class can be used to *edit* a file, so it's handy to have it in disk, so you are allow to read and write at same time. Every other storage does this, AFAIK.

Reply:
So... block blobs don't support random write/read access, so I'll scrap this and uploadwhen the file is closed (only if the file was modified).